Add Terraform and fix shell

echohack opened the pull request:
Signed-off-by: echohack <echohack@users.noreply.github.com>

This pull request adds Terraform for AWS to on-prem-builder. There are some problems with the shell scripts that prevented this from running, so I ran `shellcheck` and fixed several issues.

Furthermore, I crossed some wires and merged this pull request early, so this squashes some earlier commits to clean up the commit log.

echohack commented:
A few more notes:

`shellcheck` fixes include using underscores instead of hyphens in function names, which was causing the script to not execute at all on CentOS 7. I went ahead and fixed up many of the other issues as well and tested them.

https://echohack.app is actually run and built directly off the terraform from this branch. So if there's any question about it working successfully... there it is.

There are some minor issues around syncing upstream packages. Overall, it's probably smarter to directly people to do the package syncing on the server itself, because otherwise there's a chance that every_package_dependency will be uploaded (and then ignored because it already exists). We had a bake off on this aspect and it turned a 30 minute process into process that took 1.5 days(!!)

chefsalim commented on bldr.env.sample:
```diff
@@ -45,10 +45,10 @@ export OAUTH_TOKEN_URL=https://github.com/login/oauth/access_token
 
 # The OAUTH_CLIENT_ID is the registered OAuth2 client id
-export OAUTH_CLIENT_ID=<YOUR-CLIENT-ID>
+export OAUTH_CLIENT_ID=3333
```
Is this intentional?

chefsalim replied:
Ok, yeah something like 0 would be better..

chefsalim commented on scripts/install-hab.sh:
```diff
@@ -1,3 +1,6 @@
-type curl >/dev/null 2>&1 || { echo >&2 "curl is required for installation of habitat, but was not found. Exiting."; exit 1; }
```
Is losing the curl check intentional?

chefsalim commented on terraform/aws/main.tf:
```diff
@@ -0,0 +1,235 @@
+  }
+
+  ingress {
+    from_port   = 9631
```
Not sure we need 9631 or 9638 ingress since all the services are on the same instance..

chefsalim commented on terraform/aws/outputs.tf:
```diff
@@ -0,0 +1,10 @@
+# output "public_ips_initial_peer" {
```
Remove this file?

chefsalim commented:
Will be good to also add verbiage to the README re: the feature status, support, tshooting, etc.

predominant commented on scripts/provision.sh:
```diff
@@ -2,16 +2,9 @@
-  echo "ERROR: bldr.env file is missing!"
-  exit 1
-fi
+source ../bldr.env
```
I'd really like for the bldr.env file check to remain.

echohack commented:
@chefsalim @predominant Thank you for the feedback! I went ahead and addressed all of these issues, and went the extra mile and filled out some descriptions on the variables.

I've made it clear that this is an experimental feature in the readme as well.
